### ISSUE TITLE
Raise NoCredentialsError when no credentials are found

### ISSUE THREAD

**Issue by tehtechguy** [boto/botocore#1791]
This topic has come up before, see issues #29 and #1403. I'm re-raising this, with some details below. I have also created a PR to fix this #1752.

Certain credential providers, such as the `InstanceMetadataProvider` have the chance to fail to load credentials on start, even when the credentials exist and valid. Typically, such failures would get handled by the `RefreshableCredentials` class; however, that refresh only occurs if the provider was able to find credentials (thereby being the chosen credential provider).

The above situation results in a scenario where it's possible to create a session that does not have credentials. This will result in the client raising a `NoCredentialsError` upon the first usage of the credentials. That error will not have the ability to auto-resolve.

Having systematic credential failures like this are critical to be found upon session creation, rather than client use, as it enables services utilizing the session to know with certainty whether the session is usable or not.

To expand on this further:

- The current implementation is different from the Java SDK. In Java an exception is thrown when no credentials are found in the chain. See [this](https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-core/src/main/java/com/amazonaws/auth/AWSCredentialsProviderChain.java#L136).
- This has the potential for a web service to start, thinking it is healthy, pass a health check, and yet be unable to access any AWS services. A potential workaround is to call `session.get_credentials()` and verify that it is not `None`; however, this places an unnecessary burden on the user, since that workaround would need to be implemented by all endpoints.
- Similar to the above - any background workers will also have the same issue. That problem becomes further complicated, because there could be multiple calls, by the same application, to the instance metadata service where some calls passed and others didn't.
- Imagine a containerized implementation, such as Kubernetes, where there is the potential to have mutiple containers running on the same node, where some are able to reach the instance store and others are not. This is worsened when there any many containers running on one node (think lots of lightweight services), especially upon initial deploy, where the [thundering herd problem](https://en.wikipedia.org/wiki/Thundering_herd_problem) may become an issue.
- For anyone expecting a provider that will refresh credentials, this will never occur, because no instance of `RefreshableCredentials` will be created (the credentials will be `None`).

**Comment by tehtechguy** [comment]
PR updated to reflect the current state of the `develop` branch + all broken tests fixed.

**Comment by Krzysztof-Szyper-Epam** [comment]
It would be nice to have this fixed some day. We hit NoCredentialsError almost every single day.
Hope AWS will have a look at your PR and make something good with it.

**Comment by chris-erickson** [comment]
Can someone shed some more context on how this happens?  We have sporadic `NoCredentialsError` with a service running on EKS that uses an IAM Role to access Dynamo.  It's super annoying and impossible to predict or manage.  Is there some awful workaround that people are using (manually authenticating?  changing retry env var settings for the metadata service?)?  This seems shockingly catastrophic so I'm hoping I just don't fully understand this issue yet and have maybe misdiagnosed or problem.

**Comment by Krzysztof-Szyper-Epam** [comment]
We are using retry and it works now. But it's shameful for AWS to have such breaking bug for so long that can be easily solved internally.

**Comment by chris-erickson** [comment]
What do you mean by using retry?

**Comment by Krzysztof-Szyper-Epam** [comment]
Usually one retry is enough, but you can set it to for example 3.
Run problematic query in a loop using try-catch for NoCredentialsError exception. Use 1 second sleep inside it to make sure AWS API set new credentials properly.
Fail only after all tries are exhausted.

**Comment by chris-erickson** [comment]
Oof.  That's pretty awful when you have dozens of DynamoDB calls in a single app when any one of them could fail ☹️

We've upped the retries and timeout (via env vars) when using the metadata service, so I'm hoping that also helps.  Our other less bad option is to have a specific user for these instances that otherwise has the same roles as the nodes themselves would have and use the API key.  Tragic regression in best practices, but not working at all is not a great option either 😅

**Comment by itajaja** [comment]
I am also getting this on EKS. I am (pretty consistently) able to reproduce it:

```sh
kubectl run --restart='Never' --rm -it btest --image=ombu/boto3:1.8 -- python -c "import time; time.sleep(0); import boto3; boto3.client('s3').download_file('foo', 'bar', 'baz')"
```

this basically runs an image with bot and tries to download a file. most of the times, I get `botocore.exceptions.NoCredentialsError: Unable to locate credentials`.

Now, if I slightly modify my script to wait a little before running the s3 command:

```sh
kubectl run --restart='Never' --rm -it btest --image=ombu/boto3:1.8 -- python -c "import time; time.sleep(10); import boto3; boto3.client('s3').download_file('foo', 'bar', 'baz')"
```

most of the times (but not always!) the call succeeds. so, it seems related to some sort of lazy loading of the credentials that are not present when the container starts, or something along these lines.

I am running on EKS, kubernetes version `1.14`. also running calico (not sure if that makes a difference).

**Comment by itajaja** [comment]
I think I am finding some useful logs:

```
botocore.exceptions.ConnectTimeoutError: Connect timeout on endpoint URL: "http://169.254.169.254/latest/meta-data/iam/security-credentials/"
2020-01-23 03:39:24,361 botocore.utils [DEBUG] Max number of attempts exceeded (1) when attempting to retrieve data from metadata service.
```

really weird that it's a timeout error. how short can that timeout be? Interestengly running this command seems to reproduce it:

```
k run --restart='Never' --rm -it btest --image=python -- wget http://169.254.169.254/latest/meta-data/iam/security-credentials/ -T 1
```

yields

```
failed: Connection timed out.
Retrying.

--2020-01-23 03:49:35--  (try: 2)  http://169.254.169.254/latest/meta-data/iam/security-credentials/
Connecting to 169.254.169.254:80... connected.
HTTP request sent, awaiting response... 200 OK
```

so, works at second try. I don't think this is an issue with boto proper, but boto could have a small retry mechanism there

**Comment by chris-erickson** [comment]
Interesting intel.  This reminds me that most of our issues with Dynamo creds have gone away since introducing the timeout/retry env vars.

**Comment by bhtucker** [comment]
This is still an issue. I wonder if it's exaggerated on EKS, where the additional network hops may eat into an otherwise-reasonable timeout.

For anyone reading the above, the env vars are AWS_METADATA_SERVICE_TIMEOUT and AWS_METADATA_SERVICE_NUM_ATTEMPTS, and these settings _do_ apply to the case where IMDS returns a 200 with an empty body (ie, botocore does see that as worth retrying).

**Comment by itajaja** [comment]
that seems to help, thanks @bhtucker

**Comment by attilasimon72** [comment]
We have just worked around this bug by setting:

AWS_METADATA_SERVICE_TIMEOUT=65 (or any reasonably high value)
AWS_METADATA_SERVICE_NUM_ATTEMPTS=1

Due to #2304, this config:

AWS_METADATA_SERVICE_TIMEOUT=1
AWS_METADATA_SERVICE_NUM_ATTEMPTS=15

is practically equivalent with:

AWS_METADATA_SERVICE_TIMEOUT=1
AWS_METADATA_SERVICE_NUM_ATTEMPTS=1

(If the IMDS query runs on `ReadTimeoutError` then botocore ignores AWS_METADATA_SERVICE_NUM_ATTEMPTS and breaks the retry loop on the first try.)

If the IMDS query ends with timeout then botocore will cache the empty credential as if it was valid. Setting AWS_METADATA_SERVICE_TIMEOUT to a high value reduces the chance of getting/caching an empty cred.

**Comment by stobrien89** [comment]
Hi all,

Our apologies for the lack of attention to this.  We're unable to make this the default behavior, because not all functionality on a boto3 client requires credentials— this would be a breaking change for some customers.  

Calling `get_credentials` on a session and verifying that it is not `None` is the best way I'm aware of to force credential resolution out-of-the-box, but it seems like the biggest culprit of 'missing' credentials is IMDS throttling.  We've seen customers adopt custom rollout strategies to avoid this. We've also seen customers write custom credential providers that grab credentials from IMDS and proxy them to clients that request them in order to reduce the number of calls made to IMDS.

Like @bhtucker mentioned, the additional network hops do come into play (especially for IMDSv2) in containerized environments.  You may need a larger hop limit (default is 1) for the initial PUT request that fetches the metadata token.  More information on how to configure IMDSv2 settings can be found [here](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-instance-metadata-service.html).  

All of that said, we're willing to look into solutions that would make credential providers more resilient if this is still a widespread issue.  Feel free to re-raise an issue with us if you're still experiencing problems— we're just unable to accept this particular change.

**Comment by tehtechguy** [comment]
Thanks for coming back to this. Not merging, due to breaking changes is understood. That being said, this is still an issue. We encountered this with our own k8s install as well as with EKS. We've done the following things to work around it:

1. Set `AWS_METADATA_SERVICE_NUM_ATTEMPTS=3`
2. Pre-initialize a boto session on the class. This is a simple loop that checks `boto_session.get_credentials()`. It's configured to attempt 25 times (waiting a second between each attempt), to enable as many attempts as possible before the first k8s ready check (occurs at 30 seconds).
3. Set the ready check to be a function of the existence of a valid boto session. In this case, if `self.boto_session.get_credentials() is None` it throws an exception and k8s will nuke the pod.
4. Same as in (3) except for the k8s health check.
6. Increased IMDS session duration

Doing all of the above largely solved the problem, but it's not perfect. In theory, the pod won't come up until a boto session with valid credentials exists. It doesn't solve the thundering herd problem, but it does at least ensure that your service isn't marked as healthy until credentials are loaded. It also doesn't address the credential refresh problem and challenges can still exist for applications leveraging concurrency.

Our situation was further exacerbated by the very high request rate and our stringent latency requirements. This meant that we were additionally hurt when the credentials needed to be refreshed (ideally this would be proactive and occur in a background thread).

As an AWS customer, you're essentially forced to push your own strategy to solve this (proxy service, stop using IMDS and refresh credentials manually, etc...). This is a high friction point for something that should "just work". Ideally IMDS would be built to handle this. I can't imagine that it's a high load problem, given that the checks occur on the same instance and that an in-memory cache could be leveraged, resulting in what should be low µs response times.

With all of that being said, boto could do some more to help with this, but I'll concede that this problem is beyond boto. Hopefully some of the strategies enumerated in the post will help others.

**Comment by ChristophShyper** [comment]
If it cannot be added as default behaviour then maybe add additional parameter that will run this check?

**Comment by stobrien89** [comment]
Hi @tehtechguy and @ChristophShyper,

Thanks for the understanding and all of the helpful information you've provided so far.  I've spoken to the team about what we'd be able to do to help and we're open to possibly introducing a flag/opt-in to make `ReadTimeoutError`s retry-able.  We're still interested in hearing what would be most helpful in alleviating some of the friction here. We can take the conversation over to #2304 if everyone is agreeable to that.
